### ISSUE TITLE
How to prepare data/images.t7

### ISSUE THREAD

**Issue by EvaFlower** [saquil/RankCGAN#1]
Hi there,

Would you please explain how to prepare data/images.t7? What are its type and size? BTW, what is the type and size of shoe_rank.t7 and rank_t.t7?

Thanks!

**Comment by saquil** [comment]
Hi there, 

Thanks for letting me know that the dataset specification is missing. I was into torch before pytorch emerges. 

You do not have to stick with t7 files format, you could alter the code to use any format you are comfortable with (numpy, pickle, mat...)

As below the data specs used in my code:

images.t7:  [dataset size, channels, height size, width size],  in the code case: [50025, 3, 64, 64]

shoe_rank.t7: [modes (train/val), number of attributes, number of pairs, pair details], in the code case: [2, 4, depends on the attribute, 6] pretty much the same format as the mat file described in Zappos dataset.

rank_t.t7: [number of pairs, pair details], pair details = ('image 1 index', 'image 2 index', 'label'), the code case: [50000, 3]

You can find below links to download all these files.

https://drive.google.com/open?id=1GuxQz4FNG3GZldrXWjpA6Jh4wDufAWeH
https://drive.google.com/open?id=1eUOATRzTb_VSwvkLUje-Zw2bV5b7hdea
https://drive.google.com/open?id=1wGEn_Cz58RGaHcMds4BE4WLN5D54zZVo

In order to read t7 file in python you can use:

`from torch.utils.serialization import load_lua`
`data = load_lua(path_file)`

I hope that everything is clear now. 
Bests.

**Comment by EvaFlower** [comment]
@saquil 
Thanks for the clarification. Now it is clear.

I have other questions for your experiments in the paper. 
1) For PubFig and OSR dataset, you mentioned that you construct pairwise comparisons in a category level. Do you mean that, for example, the interested attribute for PubFig data is smiling, you use the binary attribute labels to construct the comparisons, i.e., x1>x2 for y1=1 and y2=0 (x is data, y is the label)? If so, do you ever try less comparisons, like 5K, can you model still achieve good results? 
The reason you did not use the true pairwise comparisons (annotators give the comparisons for pair of images.), is it because that the available amount is insufficient?

2) I have applied your method in another dataset -- LFW (http://vis-www.cs.umass.edu/lfw/). I construct the pairwise comparisons  (category level) as I mentioned above. When the image size is 64, your method can get some reasonable results. But when the image size is set to 256, then you method cannot produce meaningful images, only random noise within and after 500 training epochs. Could you please help figure out the problem? 

Regards,

**Comment by saquil** [comment]
@EvaFlower, below the answer to your questions.

1.1 In the classical scenario, the pairwise comparisons are defined in image level, where each pair of images is annotated. In the category level scenario, instead of annotating pairs of images, pairs of categories are annotated, with a group of images belonging to a specific category. For instance, in PubFig annotations for a specific attribute, having cat_1>cat_2 means that all images in cat_1 have more attribute presence than images in cat_2. Using this information, we build pairwise comparisons in image level for the model. If I remember, OSR annotations are slightly different than PubFig, because they do not provide pairwise comparisons of categories but a binary label for each category to define an attribute.

1.2 I think having less comparisons is fine as long as the comparisons can define a visually meaningful attribute that the model can learn. If I remember UT-Zap50K annotations are less than 10K.

1.3 PubFig and OSR do not provide image level pairwise comparisons only category level comparisons. That is the reason behind using category level, not because of the number of pairs.   

2.1 This method was built on top of DCGAN. If I remember, DCGAN can only generate 64x64 to 128x128 images, the size of 256 is not possible. In this case, I suggest to adapt this method to super-resolution GANs [1,2].

[1] Photo-Realistic Single Image Super-Resolution Using a Generative Adversarial Network, CVPR 2017.
[2] StackGAN: Text to Photo-realistic Image Synthesis with Stacked Generative Adversarial Networks, ICCV 2017.

Bests.
